private/pkg/transport/http/httpserver/httpserver.go

```diff
@@ -0,0 +1,180 @@
+
+// PrefixedHTTPHandlerMapperOption is an option for a new PrefixedHTTPHandlerMapper.
+type PrefixedHTTPHandlerMapperOption func(*prefixedHTTPHandlerMapper)
+
+// PrefixedHTTPHandlerMapperWithGlobalPrefix returns a new PrefixedHTTPHandlerMapperOption that uses the global prefix.
+func PrefixedHTTPHandlerMapperWithGlobalPrefix(globalPrefix string) PrefixedHTTPHandlerMapperOption {
+	return func(prefixedHTTPHandlerMapper *prefixedHTTPHandlerMapper) {
+		prefixedHTTPHandlerMapper.globalPrefix = globalPrefix
+	}
+}
+
+// PrefixedHTTPHandler is an http.Handler with a path prefix.
+//
+// A router should route all requests with the path prefix to this handler.
+type PrefixedHTTPHandler interface {
+	http.Handler
+	PathPrefix() string
+}
+
+// NewPrefixedHTTPHandler creates a new PrefixedHTTPHandler using a handler and prefix.
+func NewPrefixedHTTPHandler(handler http.Handler, prefix string) PrefixedHTTPHandler {
+	return newPrefixedHTTPHandler(handler, prefix)
+}
+
+// Runner is a runner.
+type Runner interface {
+	// Run runs the router.
+	//
+	// Blocking.
+	// The runner is cancelled when the input context is cancelled.
+	// The listener is closed upon return.
+	//
+	// Response write errors are logged. Response write errors can be ignored.
+	//
+	// Can be called multiple times, resulting in different runs.
+	Run(ctx context.Context, listener net.Listener, mappers ...Mapper) error
+}
+
+// NewRunner returns a new Runner.
+func NewRunner(logger *zap.Logger, options ...RunnerOption) Runner {
+	return newRunner(logger, options...)
+}
+
+// RunnerOption is an option for a new Runner.
+type RunnerOption func(*runner)
+
+// RunnerWithShutdownTimeout returns a new RunnerOption that uses the given shutdown timeout.
+//
+// The default is to use DefaultShutdownTimeout.
+// If shutdownTimeout is 0, no graceful shutdown will be performed.
+func RunnerWithShutdownTimeout(shutdownTimeout time.Duration) RunnerOption {
+	return func(runner *runner) {
+		runner.shutdownTimeout = shutdownTimeout
+	}
+}
+
+// RunnerWithReadHeaderTimeout returns a new RunnerOption that uses the given read header timeout.
+//
+// The default is to use DefaultReadHeaderTimeout.
+// If readHeaderTimeout is 0, no read header timeout will be used.
+func RunnerWithReadHeaderTimeout(readHeaderTimeout time.Duration) RunnerOption {
+	return func(runner *runner) {
+		runner.readHeaderTimeout = readHeaderTimeout
+	}
+}
+
+// RunnerWithTLSConfig returns a new RunnerOption that uses the given tls.Config.
+//
+// The default is to use no TLS.
+func RunnerWithTLSConfig(tlsConfig *tls.Config) RunnerOption {
+	return func(runner *runner) {
+		runner.tlsConfig = tlsConfig
+	}
+}
+
+// RunnerWithObservability returns a new RunnerOption that turns on
+// OpenCensus tracing and metrics.
+//
+// The default is to not turn on observability.
+func RunnerWithObservability(middleware func(http.Handler) http.Handler) RunnerOption {
+	return func(runner *runner) {
+		runner.observability = middleware
+	}
+}
```

Comment:
This looks like it has changed from the current internal version of this package? While this does drop the direct dependency on the observability libraries that it pulled in before, due to the way `RunnerWithObservability` is used it is now essentially an option to add arbitrary middleware here.

I think @mfridman had some ideas around how mappers/runners should take middleware, lets sync on this offline.